src/preamble.js

```diff
@@ -143,8 +143,9 @@ var cwrap, ccall;
       var ret = 0;
       if (str !== null && str !== undefined && str !== 0) { // null string
         // at most 4 bytes per UTF-8 code point, +1 for the trailing '\0'
-        ret = Runtime.stackAlloc((str.length << 2) + 1);
-        writeStringToMemory(str, ret);
+        var len = (str.length << 2) + 1;
```

Comment:
Kept this length computation as is for now. It has the advantage of being a fast constant time, but it is not actually correct. The UTF-8 code points as specced by Unicode all use at most 4 bytes, but earlier before UTF-8 was standardized, the spec had a wording that allowed encoding and decoding any uint32 value losslessly. That encoding can take 6 bytes per UTF-8 code point. This 6 byte wording was removed in standardized specs, but a number of UTF-8 libraries still support it, because it is free to do so (they're compatible), and it allows transmitting arbitrary binary data in a UTF-8 string stream. Emscripten's `stringToUTF8` and `UTF8ToString` functions also support this 6 byte encoding.

The 4 bytes per code point is quite wasteful, and could be replaced by `lengthBytesUTF8` here, but didn't change for now, since this form is constant time and `lengthBytesUTF8` is linear on the length of the string.